common/djangoapps/student/models.py

```diff
@@ -1442,6 +1444,8 @@ def enrollment_mode_for_user(cls, user, course_id):
             and is_active is whether the enrollment is active.
         Returns (None, None) if the courseenrollment record does not exist.
         """
+        if user.is_anonymous():
```

Comment:
For #2, it already returns `CourseEnrollmentState(None, None)` when the record is missing, so it seems like it should be able to handle this.